contributors/design-proposals/alternate-api-representations.md

```diff
@@ -0,0 +1,405 @@
+perform a set of common operations across a wide range of groups and versions and leverage
+a predictable schema.
+
+We also foresee increasing difficulty in building clients that must deal with extensions -
+there are at least 6 known web-ui or CLI implementations that need to display some
+information about third party resources or additional API groups registered with a server
+without requiring each of them to change.  Providing a server side implementation will
+allow clients to retrieve meaningful information for the `get` and `describe` style
+operations even for new API groups.
+
+
+## Implementation
+
+The HTTP spec and the common REST paradigm provide mechanisms for clients to [negotiate
+alternative representations of objects (RFC2616 14.1)](http://www.w3.org/Protocols/rfc2616/rfc2616.txt)
+and for the server to correctly indicate a requested mechanism was chosen via the `Accept`
+and `Content-Type` headers. This is a standard request response protocol intended to allow
+clients to request the server choose a representation to return to the client based on the
+server's capabilities. In RESTful terminology, a representation is simply a known schema that
+the client is capable of handling - common schemas are HTML, JSON, XML, or protobuf, with the
+possibility of the client and server further refining the requested output via either query
+parameters or media type parameters.
+
+In order to ensure that generic clients can properly deal with many different group versions,
+we introduce the `api.k8s.io` group with version `v1` that grandfathers all existing resources
+currently described as "unversioned".  A generic client may request that responses be applied
+in this version. The contents of a particular API group version would continue to be bound into
+other group versions (`status.v1.api.k8s.io` would be bound as `Status` into all existing
+API groups).  We would remove the `unversioned` package and properly home these resources in
+a real API group.
+
+
+### Considerations around choosing an implementation
+
+* We wish to avoid creating new resource *locations* (URLs) for existing resources
+  * New resource locations complicate access control, caching, and proxying
+  * We are still retrieving the same resource, just in an alternate representation,
+    which matches our current use of the protobuf, JSON, and YAML serializations
+  * We do not wish to alter the mechanism for authorization - a user with access
+    to a particular resource in a given namespace should be limited regardless of
+    the representation in use.
+  * Allowing "all namespaces" to be listed would require us to create "fake" resources
+    which would complicate authorization
+* We wish to support retrieving object representations in multiple schemas - JSON for
+  simple clients and Protobuf for clients concerned with efficiency.
+* Most clients will wish to retrieve a newer format, but for older servers will desire
```

Comment:
I think the challenge is something like Scale:

1. Scale has a server wide version that we version independent of a resource like a pod
2. Adding v2 pod does not mean that v2 scale is added
3. The version number in the path applies to the primary object, but all sub resources have no mechanism to control version

I think it would be unusual to say `GET /api/v1/namespace/foo/pods/bar -H "Accept: ... as v2"`.  But not unusual to say `GET /api/v1/namespace/foo/pods/bar/resources -H "Accept: as resourcesv1.Scale"`